pdata/internal/cmd/pdatagen/internal/primitive_slice_structs.go

```diff
@@ -0,0 +1,190 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal // import "go.opentelemetry.io/collector/pdata/internal/cmd/pdatagen/internal"
+
+import (
+	"os"
+	"strings"
+)
+
+const primitiveSliceTemplate = `// ${structName} represents a []${itemType} slice.
+// The instance of ${structName} can be assigned to multiple objects since it's immutable.
+//
+// Must use New${structName} function to create new instances.
+// Important: zero-initialized instance is not valid for use.
+type ${structName} internal.${structName}
+
+func (ms ${structName}) getOrig() *[]${itemType} {
+	return internal.GetOrig${structName}(internal.${structName}(ms))
+}
+
+// New${structName} creates a new empty ${structName}.
+func New${structName}() ${structName} {
+	orig := []${itemType}(nil)
+	return ${structName}(internal.New${structName}(&orig))
+}
+
+// AsRaw returns a copy of the []${itemType} slice.
+func (ms ${structName}) AsRaw() []${itemType} {
+	return copy${structName}(*ms.getOrig())
+}
+
+// FromRaw copies raw []${itemType} into the slice ${structName}.
+func (ms ${structName}) FromRaw(val []${itemType}) {
```

Comment:
@bogdandrutu do you think we should return ms here?

Otherwise current inline usages of `pcommon.NewImmutable...Slice()` becomes pretty clunky

Reply:
As for me, I'm not sure about that. I don't like returning it and I don't like how it complicates the usages

Reply:
Let me apply current state of this PR on contrib, and we'll see

Reply:
It looks pretty good https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/14016 I believe we should keep current version